sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java

```diff
@@ -0,0 +1,488 @@
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+
+      if (queue != null) {
+        builder.add(DisplayData.item("queue", queue));
+      }
+      if (topic != null) {
+        builder.add(DisplayData.item("topic", topic));
+      }
+    }
+
+    ///////////////////////////////////////////////////////////////////////////////////////
+
+    private ConnectionFactory connectionFactory;
+    private String queue;
+    private String topic;
+    private long maxNumRecords;
+    private Duration maxReadTime;
+
+    private Read() {}
+
+    private Read(
+        ConnectionFactory connectionFactory,
+        String queue,
+        String topic,
+        long maxNumRecords,
+        Duration maxReadTime) {
+      super("JmsIO.Read");
+
+      this.connectionFactory = connectionFactory;
+      this.queue = queue;
+      this.topic = topic;
+      this.maxNumRecords = maxNumRecords;
+      this.maxReadTime = maxReadTime;
+    }
+
+    /**
+     * Creates an {@link UnboundedSource<JmsRecord, ?>} with the configuration in
+     * {@link Read}. Primary use case is unit tests, should not be used in an
+     * application.
+     */
+    @VisibleForTesting
+    UnboundedSource<JmsRecord, JmsCheckpointMark> createSource() {
+      return new UnboundedJmsSource(
+          connectionFactory,
+          queue,
+          topic);
+    }
+
+  }
+
+  private JmsIO() {}
+
+  private static class UnboundedJmsSource extends UnboundedSource<JmsRecord, JmsCheckpointMark> {
+
+    private final ConnectionFactory connectionFactory;
+    private final String queue;
+    private final String topic;
+
+    public UnboundedJmsSource(
+        ConnectionFactory connectionFactory,
+        String queue,
+        String topic) {
+      this.connectionFactory = connectionFactory;
+      this.queue = queue;
+      this.topic = topic;
+    }
+
+    @Override
+    public List<UnboundedJmsSource> generateInitialSplits(
+        int desiredNumSplits, PipelineOptions options) throws Exception {
+      return Collections.singletonList(new UnboundedJmsSource(connectionFactory, queue, topic));
+    }
+
+    @Override
+    public UnboundedJmsReader createReader(PipelineOptions options,
+                                           JmsCheckpointMark checkpointMark) {
+      return new UnboundedJmsReader(this, checkpointMark);
+    }
+
+    @Override
+    public void validate() {
+      if (connectionFactory == null) {
+        throw new IllegalArgumentException("ConnectionFactory is not defined");
+      }
+      if (queue == null && topic == null) {
+        throw new IllegalArgumentException("Either queue or topic is not defined");
+      }
+    }
+
+    @Override
+    public Coder getCheckpointMarkCoder() {
+      return SerializableCoder.of(JmsCheckpointMark.class);
+    }
+
+    @Override
+    public Coder<JmsRecord> getDefaultOutputCoder() {
+      return SerializableCoder.of(JmsRecord.class);
```

Comment:
I looked briefly at JmsRecord -- it seems hard to do much better than a Serializable coder or a JSON coder if you have a `Map<String, Object>` in the body.